flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java

```diff
@@ -201,10 +221,32 @@ public ScanRuntimeProvider getScanRuntimeProvider(ScanContext context) {
         final TypeInformation<RowData> producedTypeInfo =
                 context.createTypeInformation(producedDataType);
 
-        final FlinkKafkaConsumer<RowData> kafkaConsumer =
-                createKafkaConsumer(keyDeserialization, valueDeserialization, producedTypeInfo);
+        if (fallbackToFlinkKafkaConsumer) {
+            final FlinkKafkaConsumer<RowData> kafkaConsumer =
+                    createKafkaConsumer(keyDeserialization, valueDeserialization, producedTypeInfo);
+            return SourceFunctionProvider.of(kafkaConsumer, false);
+        }
 
-        return SourceFunctionProvider.of(kafkaConsumer, false);
+        final KafkaSource<RowData> kafkaSource =
+                createKafkaSource(keyDeserialization, valueDeserialization, producedTypeInfo);
+
+        return new DataStreamScanProvider() {
```

Comment:
why can't we use `SourceProvider`?

Reply:
Currently ```SourceProvider``` doesn't support watermark pushdown, so I have to use DataStreamSourceProvider for now:

https://github.com/apache/flink/blob/bc984f41c7d18d88a438ed071f563ac0bc5bd005/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecTableSourceScan.java#L98-L103

Reply:
Actually, fixing this would have been part of the Kafka source issue as it is the first production ready streaming source with watermarks. But let's postpone it after feature freeze now.